networkx/algorithms/approximation/treewidth.py

```diff
@@ -0,0 +1,272 @@
+    return treewidth_decomp(G, MinFillInHeuristic)
+
+
+class MinDegreeHeuristic:
+    def __init__(self, graph):
+        self._graph = graph
+
+        # A collection of nodes that have to be updated in the heap before each iteration
+        self._update_nodes = []
+
+        # self._degreeq is heapq with 2-tuples (degree,node)
+        self._degreeq = []
+        # build heap with initial degrees
+        for n in graph:
+            self._degreeq.append((len(graph[n]), n))
+        heapify(self._degreeq)
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        #Implement next method for backwards compatibility with python 2.
+        return self.__next__()
+
+    def __next__(self):
+        # Update nodes in self._update_nodes
+        for n in self._update_nodes:
+            # insert changed degrees into degreeq
+            heappush(self._degreeq, (len(self._graph[n]), n))
+
+        while self._degreeq:
+            # get the next (minimum degree) node
+            (min_degree, elim_node) = heappop(self._degreeq)
+            if elim_node not in self._graph or len(self._graph[elim_node]) != min_degree:
+                # Outdated entry in degreeq
+                continue
+            elif min_degree == len(self._graph) - 1:
+                # Fully connected: Abort condition
+                raise StopIteration
+
+            # Remember to update nodes in the heap before getting the next node
+            self._update_nodes = self._graph[elim_node]
+            return elim_node
+
+        # The heap is empty: Abort
+        raise StopIteration
+
+
+class MinFillInHeuristic:
+
+    def __init__(self, graph):
+        self._graph = graph
+
+    def __iter__(self):
+        return self
+
+    def next(self):
+        #Implement next method for backwards compatibility with python 2.
+        return self.__next__()
+
+    def __next__(self):
+        if len(self._graph) == 0:
+            raise StopIteration
+
+        min_fill_in_node = None
+
+        min_fill_in = sys.maxsize
+
+        # create sorted list of (degree, node)
+        degree_list = [(len(self._graph[node]), node) for node in self._graph]
+        degree_list.sort()
+
+        # abort condition
+        min_degree = degree_list[0][0]
+        if min_degree == len(self._graph) - 1:
+            raise StopIteration
+
+        for (_, node) in degree_list:
+            num_fill_in = 0
+            # Convert to list in order to access by index
+            nbrs = list(self._graph[node])
+            for i in range(len(nbrs) - 1):
+                for j in range(i + 1, len(nbrs)):
+                    if nbrs[j] not in self._graph[nbrs[i]]:
+                        num_fill_in += 1
+                        # break inner loop if this can't be min-fill-in node anymore
+                        if num_fill_in >= min_fill_in:
+                            break
+
+                if num_fill_in >= min_fill_in: # break outer loop
+                    break
+
```

Comment:
This loop structure doesn't seem to need the individual loops. Maybe a set operation can get the number of nodes you want without the loop overhead.  I'm not sure of the exact logic -- does this or something similar work?

    for (deg, node) in degree_list:
        nbrs = graph[node]
        for nbr in nbrs:
            num_fill_in += len(nbrs - graph[nbr])
            if num_fill_in >= min_fill_in:
                break